Correcting calculation of water year for the case if wyMonth < 3

jenskiesel opened the pull request:
I updated the function "get_waterYear" for the case if a user sets wyMonth < 3. With the previous calculation, Jan and Feb of the time series were assigned the wrong water year. This caused problems with the leap years and the validate data function returned missing days even though the time series was complete. See the comments I added to the function and some further explanation and tests in this issue:
https://github.com/DOI-USGS/EflowStats/issues/196

Commit:
Corrected wrong assignment of water year for the months of Jan and Feb if the desired start of the water year is earlier than March (wyMonth < 3)

Commit:
Corrected wrong assignment of water year for the months of Jan and Feb if the desired start of the water year is earlier than March (wyMonth < 3)

Commit:
Corrected wrong assignment of water year for the months of Jan and Feb if the desired start of the water year is earlier than March (wyMonth < 3)

Commit:
Merge branch 'main' of https://github.com/jenskiesel/EflowStats into main

Commit:
Merge branch 'main' of https://github.com/jenskiesel/EflowStats into main

Commit:
Corrected the calculation of the water year to a more intuitive convention: the water year value should be the calendar year with the most dates in the water year. Therefore, a cutoff is assigned at wyMonth=7. See this issue for further information: https://github.com/DOI-USGS/EflowStats/issues/196

Commit:
Adjusted the validate_data and isleapYear functions to work with the new assignment of the water year (cutoff at wyMonth=7). See this issue for details: https://github.com/DOI-USGS/EflowStats/issues/196

jenskiesel commented:
According to @jds485's [comment,](https://github.com/DOI-USGS/EflowStats/issues/196#issuecomment-1667898933) I changed the assignment of the water year so that the water year value is the calendar year with the most dates in the water year. This means, that the month of July is the cutoff and Jan-Jun get assigned the previous year's value and Jul-Dec get assigned the current year's value.

Changing this convention causes the validate_data function to fail due to leap year issues: If the user's wyMonth is Mar, Apr or Jun, the leap day of a leap year is part of the 'previous' year. 

Example: 
Time series: 01Apr1999 - 31Mar2000
wyMonth = 4
Assigned water year: 1999
Number of days: 366 (because it contains 29Feb2000)
-> Function must return true for 1999 (and any other year that precedes a leap year)

The updated functions can be tested with:

```
test_water_year_assignment <- function(x){
  yearType <- 'water'
  for (wy_month in 1:12){
    x_cut <- cut_dataToWaterYear(x, wy_month)
    cat("wyMonth =", wy_month, ":\n")
    print(as.data.frame(table(get_waterYear(x_cut$date, wy_month))))
    cat("----------------\n")
    if (length(validate_data(x_cut, yearType, wy_month))>1){
      cat("Data is valid\n\n\n")
    }else{
      cat("Data is NOT valid\n\n\n")
    }
  }
}


# create a dummy dataset with a valid time series
x_val <- data.frame(
  date = seq.Date(from = as.Date('1990-01-01'), to = as.Date('2010-12-31'), by = 'days'),
  discharge = seq.int(1,length(seq.Date(from = as.Date('1990-01-01'), to = as.Date('2010-12-31'), by = 'days')))
)
# create a dummy dataset with an invalid time series
x_not_val <- x_val[x_val$date != as.Date('1992-02-29'),]

# assign water year and validate the data
test_water_year_assignment(x_val)
test_water_year_assignment(x_not_val)
```

Let me know if this requires further clarification or if further tests need to be conducted.

jds485 reviewed (changes requested):
Thanks! This looks good to me. I have a few small comments to address. 

@dblodgett-usgs are tests automatically run for PRs?

jenskiesel committed:
clarify description in comment

Co-authored-by: Jared D. Smith <jared.d.smith485@gmail.com>

jenskiesel committed:
clarify description in comment

Co-authored-by: Jared D. Smith <jared.d.smith485@gmail.com>

jenskiesel committed:
clarify description in comment

Co-authored-by: Jared D. Smith <jared.d.smith485@gmail.com>

jenskiesel committed:
clarify description in comment

Co-authored-by: Jared D. Smith <jared.d.smith485@gmail.com>

jenskiesel committed:
clarify description in comment

Co-authored-by: Jared D. Smith <jared.d.smith485@gmail.com>

jenskiesel committed:
clarify description in comment

Co-authored-by: Jared D. Smith <jared.d.smith485@gmail.com>

Commit:
suggestion for updating the calc_timingAverage.R function to work with the dynamic assignment of the water year and leap-year issue

Commit:
Merge branch 'main' of https://github.com/jenskiesel/EflowStats into main

jenskiesel commented on R/isleapYear.R:
```diff
@@ -1,5 +1,22 @@
-is.leapyear <- function(year){
-        #http://en.wikipedia.org/wiki/Leap_year
-        return(((year %% 4 == 0) & (year %% 100 != 0)) | (year %% 400 == 0))
+is.leapyear <- function(year, yearType='calendar', wyMonth=10L){
```
I have pushed a suggestion for updating the [calc_timingAverage.R](https://github.com/DOI-USGS/EflowStats/pull/197/commits/18d1886336e18428169f24fbccaab1a9b875a5c4) function. Let me know what you think.

dblodgett-usgs commented:
It looks like I have to click approve to get the actions to run. They are going now.

jenskiesel commented:
As far as I understand this, the checks have failed because the result from the function calc_timingAverage is different now. This could be due to the one bug I found in the function (calculating the 29th day of the year and not of the month)?

dblodgett-usgs commented:
We'll need to update tests that are failing to their new expected output values.

jds485 commented on R/calc_timingAverage.R:
```diff
@@ -57,17 +57,31 @@ calc_timingAverage <- function(x,yearType = "water",wyMonth=10L,digits=3,pref="m
+        # after removing the 29th of Feb, we can count the rows sequentially 
+        # for each unique water year
+        # https://stackoverflow.com/a/46613159
+        x$day <- sequence(rle(x$year_val)$lengths)
```
This assumes that x is ordered by date, right? Is that guaranteed? This is a much cleaner method than before!

jenskiesel replied:
Yes, you are right, it must be ordered. I think we can assume that, as the validate_data function orders by date:
```
###Order by date
        x = dplyr::arrange(x, date)
```
and validate_data is run at the beginning of the calc_timingAverage function.

jds485 replied:
Thanks! These edits look good to me, then

jenskiesel replied:
Ok, I removed the original calculation method and cleaned up the comments. 
Thanks so much @jds485 for your fast and thorough review. And thanks @dblodgett-usgs for considering this for the main package :-)!

Commit:
removed the original calculation and clenaed up commenting

dblodgett-usgs commented:
Once we have tests passing, I'm happy to go ahead and merge this if you both are happy with it.

dblodgett-usgs commented:
@jenskiesel -- can you suggest updated values for the tests that are failing so they pass then?

dblodgett-usgs commented:
Just a reminder, @jenskiesel, can you please fix the failing tests so we can get this merged?

jenskiesel commented:
Hi @dblodgett-usgs, sorry for the delay, I have had a very busy week. I will look into the tests and try to make a suggestion how to fix the values - hopefully some time early next week.

Commit:
The calculation of the leap year must also be updated in the cut_dataToWaterYear function as otherwise data is cut even though it is complete

Commit:
NA need to be removed after calculating rolling means (see also function calc_durationHigh.R where na.rm=TRUE when calculating rolling means)

jenskiesel commented:
While testing all this, I found two additional issues:

1. the "cut_dataToWaterYear" function must use the 'new' assignment of leap years as well, otherwise years are cut from the data if a water year is at the beginning/end of the time series. I removed the lubridate is leap year function and replaced by the custom one

2. This is unrelated to the leap year issue: I noticed that na.rm=TRUE is missing from the rolling mean calculations in the function "minRollingMean" in "calc_durationLow.R". Without na.rm=TRUE I had "NA" values returned for indicators "dl2.5". Also, na.rm=TRUE is already set in the "maxRollingMean" in "calc_durationHigh.R", so I think it also needs to be added to "minRollingMean".

Would be great if someone could check/confirm these changes. 
I will then try to figure out how to update the tests.

Commit:
Added yearType to \item

Commit:
added "water" (=yearType) to the test function so that cut_dataToWaterYear does not fail

jenskiesel commented:
I looked into updating the tests and the test data. For the "cut_dataToWaterYear" test we only need to pass 'water' to the function.

Sorry for my lack of experience with this, but I don't know how updating the tests is usually done for the remaining functions. I wrote a function that compares the 'old' test data to the 'new' results, shows the value differences and writes the 'new' test data to an .rds file:

```
library(dplyr)

# this function compares the new data to the old,
# prints the differences and saves the new data to an .rds file
# this .rds file can be used to replace the 'old' .rds file
evaluate_diff_and_save <- function(old_df, new_df, fill_value, output_fp) {
  old_df['statistic_new'] <- new_df['statistic']
  old_df <- old_df %>% mutate_all(~ifelse(is.na(.x) | is.nan(.x) | is.infinite(.x), -999999, .x))
  old_df$diff = ifelse(old_df$statistic == old_df$statistic_new, FALSE, TRUE)
  print(old_df[old_df$diff == TRUE, ])
  save(new_df, file=output_fp)
}
fill_value <- -999999  # used to substitute NA for difference comparison
```

When calculating all indicators with:

```
# calc all mean
data_fn <- "tests_calc_allHIT.rds"
new_df <- calc_allHIT(x=sampleData[c("date","discharge")],
            yearType="water", stats="all", pref="mean", drainArea=50)
old_df <- readRDS(paste("data", data_fn, sep="/"))
evaluate_diff_and_save(old_df, new_df, fill_value, paste("updated_data", data_fn, sep="/"))
```
I think the results make sense - the 'new' results have no NA values for the "dl" indicators and apart from 'ta2' the values agree with the test data:

```
    indice  statistic statistic_new diff
110    dl2 -999999.00        40.667 TRUE
111    dl3 -999999.00        41.214 TRUE
112    dl4 -999999.00        50.067 TRUE
113    dl5 -999999.00        59.117 TRUE
115    dl7 -999999.00        10.433 TRUE
116    dl8 -999999.00        11.029 TRUE
117    dl9 -999999.00         8.097 TRUE
118   dl10 -999999.00         0.253 TRUE
120   dl12 -999999.00         0.343 TRUE
121   dl13 -999999.00         0.417 TRUE
154    ta2      87.03        87.012 TRUE
```

I however feel reluctant updating the .rds files because I struggle to understand some of your test functions. E.g. "tests_calc_durationLow.R" contains a calculation for pref="mean" and pref="median" but both are compared to the same file "tests_calc_durationLow.rds". Also, for the "nwis" files, the results look quite different:

```
data_fn <- "tests_calc_durationLow_nwis_mean.rds"
new_df <- calc_durationLow(x=sampleData[c("date","discharge")],yearType="water",pref = "mean")
old_df <- readRDS(paste("data", data_fn, sep="/"))
evaluate_diff_and_save(old_df, new_df, fill_value, paste("updated_data", data_fn, sep="/"))
```
Results in:
```
   indice statistic statistic_new diff
1     dl1   146.000        40.000 TRUE
2     dl2   136.222        40.667 TRUE
3     dl3   139.429        41.214 TRUE
4     dl4   172.711        50.067 TRUE
5     dl5   205.107        59.117 TRUE
6     dl6    28.082        10.607 TRUE
7     dl7    15.313        10.433 TRUE
8     dl8    16.266        11.029 TRUE
9     dl9    24.957         8.097 TRUE
10   dl10    24.294         0.253 TRUE
11   dl11     0.293         0.333 TRUE
12   dl12     0.280         0.343 TRUE
13   dl13     0.347         0.417 TRUE
14   dl14     0.582         0.650 TRUE
15   dl15     0.382         0.418 TRUE
16   dl16     9.818        10.924 TRUE
17   dl17    20.678        47.920 TRUE
```

What is the best way to move forward with this?

dblodgett-usgs commented:
Hmmm... I'll have to look into it. Not really "my" method as I took over stewardship of this from someone else. I'll take a quick look and see if I can make sense of it and will have to come back to it tonight.

jenskiesel commented:
Thanks @dblodgett-usgs for the explanation! I was not aware I could simply delete the prior .rds files and that the tests create new files. 

I think we are good for most tests. But I am concerned about the "NWIS" files. If you think this is not an issue, should I just remove the failing .rds files from my fork, re-run the tests and push the resulting .rds files?

dblodgett-usgs commented:
I see... more NaNs in the new code for sample data and not that different results for NWIS.

```r
#> for sampleData in "duration low pref mean" I see:

> bind_cols(calc_durationLowTest, select(calc_durationLowTestCheck, newstat = statistic))
   indice statistic newstat
1     dl1    40.000  40.000
2     dl2    40.667     NaN
3     dl3    41.214     NaN
4     dl4    50.067     NaN
5     dl5    59.117     NaN
6     dl6    10.607  10.607
7     dl7    10.433      NA
8     dl8    11.029      NA
9     dl9     8.097      NA
10   dl10     0.253      NA
11   dl11     0.333   0.333
12   dl12     0.343     NaN
13   dl13     0.417     NaN
14   dl14     0.650   0.650
15   dl15     0.418   0.418
16   dl16    10.924  10.924
17   dl17    47.920  47.920
18   dl18     0.000   0.000
19   dl19       NaN     NaN
20   dl20     0.000   0.000

#> for sample_nwis_data.rds in "duration low pref mean" I see 

> bind_cols(calc_durationLowTest, select(calc_durationLowTestCheck, newstat = statistic))
   indice statistic newstat
1     dl1   146.000 146.000
2     dl2   147.800 136.222
3     dl3   151.171 139.429
4     dl4   177.467 172.711
5     dl5   257.564 205.107
6     dl6    28.082  28.082
7     dl7    28.119  15.313
8     dl8    27.656  16.266
9     dl9    26.658  24.957
10   dl10    32.510  24.294
11   dl11     0.293   0.293
12   dl12     0.304   0.280
13   dl13     0.356   0.347
14   dl14     0.582   0.582
15   dl15     0.382   0.382
16   dl16     9.818   9.818
17   dl17    20.678  20.678
18   dl18     0.000   0.000
19   dl19       NaN     NaN
20   dl20     0.000   0.000

#> for sampleData in "duration low pref median" I see:

> bind_cols(calc_durationLowTest, select(calc_durationLowTestCheck, newstat = statistic))
   indice statistic newstat
1     dl1    40.000  40.000
2     dl2    40.667     NaN
3     dl3    41.214     NaN
4     dl4    50.067     NaN
5     dl5    59.117     NaN
6     dl6    10.607  10.607
7     dl7    10.433      NA
8     dl8    11.029      NA
9     dl9     8.097      NA
10   dl10     0.253      NA
11   dl11     0.333   0.333
12   dl12     0.343     NaN
13   dl13     0.417     NaN
14   dl14     0.650   0.650
15   dl15     0.418   0.418
16   dl16    10.924  10.924
17   dl17    47.920  47.920
18   dl18     0.000   0.000
19   dl19       NaN     NaN
20   dl20     0.000   0.000

#> for sample_nwis_data.rds in "duration low pref median" I see:

> bind_cols(calc_durationLowTest, select(calc_durationLowTestCheck, newstat = statistic))
   indice statistic newstat
1     dl1   126.000 126.000
2     dl2   127.667 127.667
3     dl3   129.143 129.143
4     dl4   162.300 162.300
5     dl5   262.644 176.544
6     dl6    28.082  28.082
7     dl7    28.119  15.313
8     dl8    27.656  16.266
9     dl9    26.658  24.957
10   dl10    32.510  24.294
11   dl11     0.293   0.293
12   dl12     0.259   0.259
13   dl13     0.326   0.326
14   dl14     0.582   0.582
15   dl15     0.382   0.382
16   dl16     9.818   9.818
17   dl17    20.678  20.678
18   dl18     0.000   0.000
19   dl19       NaN     NaN
20   dl20     0.000   0.000
```

jenskiesel commented:
Thanks @dblodgett-usgs, I can (almost) reproduce these numbers. I think you have accidentially assigned the 'old' test statistics to the 'newstat' column. The values you have in the 'statistic' column, are the ones that are calculated from the new code. E.g. when I run the code with my changes:
```
> calc_durationLow(x=sampleData[c("date","discharge")],yearType="water",pref = "mean")
   indice statistic
1     dl1    40.000
2     dl2    40.667
3     dl3    41.214
4     dl4    50.067
5     dl5    59.117
6     dl6    10.607
7     dl7    10.433
8     dl8    11.029
9     dl9     8.097
10   dl10     0.253
11   dl11     0.333
12   dl12     0.343
13   dl13     0.417
14   dl14     0.650
15   dl15     0.418
16   dl16    10.924
17   dl17    47.920
18   dl18     0.000
19   dl19       NaN
20   dl20     0.000
```
I have only one NaN and those are the values that are labelled "statistic" in your example, but those should be the "newstat" in your example if I understand your intention correctly.

For the "nwis" data - sorry, I initially had used the "sampleData", not the "sample_nwis_data.rds". Now I get results that are closer to the old values:
```
   indice statistic statistic_new
1     dl1   146.000       146.000
2     dl2   136.222       147.800
3     dl3   139.429       151.171
4     dl4   172.711       177.467
5     dl5   205.107       257.564
6     dl6    28.082        28.082
7     dl7    15.313        28.119
8     dl8    16.266        27.656
9     dl9    24.957        26.658
10   dl10    24.294        32.510
11   dl11     0.293         0.293
12   dl12     0.280         0.304
13   dl13     0.347         0.356
14   dl14     0.582         0.582
15   dl15     0.382         0.382
16   dl16     9.818         9.818
17   dl17    20.678        20.678
18   dl18     0.000         0.000
19   dl19        NA            NA
20   dl20     0.000         0.000
```

dblodgett-usgs commented:
OK -- yeah, I was working fast and loose so not surprised I got something switched. 

Glad the NWIS test numbers are not as different as we had thought. Are those changes expected? If so, please go ahead and update the test files.

Commit:
Needed to change test functions to "expect_equal_to_reference" as the target test files are not created with "expect_equal"; readRDS functions do not work when .rds files don't exist.

Commit:
New .rds files after the changes made to the water year calculation and fixing bug that prevented the calculation of somme duration_low indicators.

jenskiesel commented:
The differences occur for indicators that were subject to the bug we fixed (not ignoring the NAs for the rolling mean). I don't know how those indicators were calculated previously and also why the tests did not fail with the old code (should have been NA). So, bottom line - I figured the differences are ok and I updated the test data that were failing. I had to change a few test functions as otherwise they did not create new files. It seems the tests now worked. 

But there is still one warning and note related to the documentation of the 'cut_dataToWaterYear' function because I had to add the variable "yearType". I don't know how to add that to the docs.

jenskiesel commented:
Thanks so much @dblodgett-usgs  and @jds485 for the support in getting this done!!
